### PR TITLE
feat: add service-worker

### LINKED ISSUE #1545: Service worker - request filtering
### Overview

We need to be able to filter and sanitize outgoing requests for Pendo to ensure security.

Request filtering is required to implement Pendo.

### References and additional details

This is purely concerned with creating a service worker capable of running a request through some filter middleware, but is not concerned with the actually middleware/filters themselves.

### Acceptance Criteria

A service worker exists that can take pluggable filters.

### Need By Date

4/25

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by cjthompson**
## Description

feat: add service-worker

`create-react-app` already includes `workbox` for supporting service-workers for PWAs, so no change to the webpack config was required.

However, because `react-scripts` automatically calls `InjectManifest`, the `service-worker` file must include a reference to `__WB_MANIFEST` or the build will fail.  That manifest can be used for doing caching of assets but we're not doing that right now.

## Notice

- [X] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [X] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #1545

## Risk

Low. The worst case is the service worker doesn't load but that won't cause anything else to stop working. It's possible there's a slight performance decrease when there are a lot of HTTP calls due to the logging of all service worker events.

## Testing

It's best to test this is an ephemeral environment.
1. Check the Console tab in Developer Tools.
2. Look for `ServiceWorker` log entries

## Screenshots (if applicable)
image

### From the Fleek build:
image

### CODE REVIEW

**Comment by 0xdef1cafe** [comment]
@cjthompson are we able to push a "breaking change" to this branch to ensure we don't cache the wrong thing and get users in a broken state? 

https://create-react-app.dev/docs/making-a-progressive-web-app/#why-opt-in

https://github.com/facebook/create-react-app/issues/2398

**Comment by cjthompson** [comment]
> @cjthompson are we able to push a "breaking change" to this branch to ensure we don't cache the wrong thing and get users in a broken state?
> 
> https://create-react-app.dev/docs/making-a-progressive-web-app/#why-opt-in
> 
> [facebook/create-react-app#2398](https://github.com/facebook/create-react-app/issues/2398)

There service-worker doesn't actual do anything right now. It doesn't cache anything so it can't cache the wrong things.

File: src/serviceWorkerRegistration.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  // @TODO: process.env.NODE_ENV === 'production'
+  if ('serviceWorker' in navigator) {
+    const publicUrl = new URL(process.env.PUBLIC_URL, window.location.href)
+
+    if (publicUrl.origin !== window.location.origin) {
+      // Our service worker won't work if PUBLIC_URL is on a different origin
+      // from what our page is served on. This might happen if a CDN is used to
+      // serve assets; see https://github.com/facebook/create-react-app/issues/2374
+      return
+    }
```
Comment (0xdef1cafe) [inline src/serviceWorkerRegistration.ts:96]: we recently removed `PUBLIC_URL` from env vars in production - will this work?

File: src/serviceWorkerRegistration.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    window.addEventListener('load', () => {
+      const swUrl = `${process.env.PUBLIC_URL}/service-worker.js`
+      debugger
```
Comment (0xdef1cafe) [inline src/serviceWorkerRegistration.ts:100]: erroneous debugger
```suggestion
```

File: src/serviceWorkerRegistration.ts (hunk 1 above)
Comment (0xdef1cafe) [reply src/serviceWorkerRegistration.ts:96]: just checked this in the console and indeed it does, nvm

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]
small requested change, and me rambling out loud

**Review by 0xdef1cafe — APPROVED** [review]

**Review by asamere — APPROVED** [review]

**Pull request merged**